File: AdventureOfCode2020/SecondOfDecemberTaskTwo/PasswordOption.cs
using System;
using System.Linq;
using System.Text.RegularExpressions;

namespace SecondOfDecemberTaskTwo
{
    public class PasswordOption
    {
        //static regex objects to prevent compiling the regex patterns multiple times
       // private static readonly Regex OccurrenceRegex = new Regex(@"([0-9]+)(?=-)");
        private static readonly Regex OccurrenceRegex = new Regex(@"([0-9]+)");
        private static readonly Regex PassCharRegex = new Regex(@"([a-z]+)(?=:)");
        private static readonly Regex PasswordRegex = new Regex(@"([a-z]+)(?!\\w)(?!:)");
        
        public int MinOccurrence { get; private set; }
        public int MaxOccurrence { get; private set; }
        public char PassChar { get; private set; }
        public String Password { get; private set; }
        public int PassCharCount { get; private set; }
        public bool PasswordValid { get; private set; }

        public PasswordOption(String PassLine)
        {
            SetMinMaxOccurrence(PassLine);
            SetPassChar(PassLine);
            SetPassword(PassLine);

            if ((MinOccurrence + "-" + MaxOccurrence + " " + PassChar + ": " + Password) != PassLine)
            {
                throw new Exception("Parsing error");
            }
            
            PasswordValid = (Password[MinOccurrence - 1] == PassChar && Password[MaxOccurrence - 1] != PassChar) ||
                            (Password[MinOccurrence - 1] != PassChar && Password[MaxOccurrence - 1] == PassChar);
        }

        private void SetMinMaxOccurrence(String input)
        {
            var minOutput = OccurrenceRegex.Matches(input)[0].Value;
            var maxOutput = OccurrenceRegex.Matches(input)[1].Value;
            MinOccurrence = Int32.Parse(minOutput);
            MaxOccurrence = Int32.Parse(maxOutput);
        }

        private void SetPassChar(String input)
        {
            var output = PassCharRegex.Match(input).Value;
            PassChar = char.Parse(output);
        }

        private void SetPassword(String input)
        {
            var output = PasswordRegex.Match(input).Value;
            Password = output;
        }
    }
}

File: AdventOfCode2020_cpp/Day2/Day2_Task1/PasswordOption.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <regex>
//
// Created by aublue on 05.12.20.
//

#include "PasswordOption.h"

PasswordOption::PasswordOption(string s) {
    _SetMinMaxOccurrence(s);
    _SetPassChar(s);
    _SetPassword(s);

    this->PassCharCount = std::count(this->Password.begin(), this->Password.end(), this->PassChar);
    this->PasswordValid = (this->MinOccurrence <= this->PassCharCount && this->PassCharCount <= this->MaxOccurrence);
}

void PasswordOption::_SetMinMaxOccurrence(string input) {

    vector<string> matches = _GetAllMatches(this->OccurrenceRegex, input);

    if(matches.size() == 2)
    {
        string min = matches[0];
        string max = matches[1];

        this->MinOccurrence = stoi(min);
        this->MaxOccurrence = stoi(max);
    }else{
        throw "Error while parsing input line: " + input;
    }
}

void PasswordOption::_SetPassChar(string input)
{
    vector<string> matches = _GetAllMatches(this->PassCharRegex, input);

    if(matches.size() == 1)
    {
        string charStr = matches[0];
        this->PassChar = charStr[0];
    }else{
        throw "Error while parsing input line: " + input;
    }
}

void PasswordOption::_SetPassword(string input)
{
    vector<string> matches = _GetAllMatches(this->PasswordRegex, input);

    if(matches.size() == 1)
    {
        this->Password = matches[0];
    }else{
        throw "Error while parsing input line: " + input;
    }
}

std::vector<string> PasswordOption::_GetAllMatches(regex reg, string inputstr)
{
    vector<string> output;
    smatch regExMatch;

    string::const_iterator searchStart (inputstr.cbegin());
    while(regex_search(searchStart, inputstr.cend(), regExMatch, reg))
    {
        output.push_back(regExMatch[0]);
        searchStart = regExMatch.suffix().first;
    }
    return output;
}